src/gamepad/linux/linux-raw-gamepad-monitor.vala

```diff
@@ -0,0 +1,102 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.LinuxRawGamepadMonitor : Object, RawGamepadMonitor {
+	public delegate void RawGamepadCallback (RawGamepad raw_gamepad);
+
+	private static LinuxRawGamepadMonitor instance;
+
+	private GUdev.Client client;
+	private HashTable<string, RawGamepad> raw_gamepads;
+
+	private LinuxRawGamepadMonitor () {
+		client = new GUdev.Client ({"input"});
+		client.uevent.connect (handle_udev_client_callback);
+
+		// Initialize internally plugged in gamepads
+		client.query_by_subsystem ("input").foreach (initial_device_iterator);
```

Comment:
First get the list, then loop through it with a regular foreach, this way you could move the code from the callback back here, using continue instead of return.